fix(service): пофикшено добавление сумм в сделку

sergeJAVA opened the pull request.

sergeJAVA committed:
fix(service): пофикшено добавление сумм в сделку

sergeJAVA committed:
feat(service): добавил генерацию xlsx файла по эндопоинту POST deal/search/export, который ответом отдаёт файл

sergeJAVA commented:
![Screenshot 2025-07-01 181748](https://github.com/user-attachments/assets/494a45a6-9292-4885-8f24-3d06994e5fd6)

получилось вот так

AlKl1M reviewed (changes requested):
Замечания в https://github.com/sergeJAVA/deal-service/commit/cf6a3f248bbabfebcc7781aa183ca104ee3ccfc7#comments т.к. у тебя в ПРе изменения неаппрувнутые, которые как я понял уже влиты в main, не делай так.

В целом код грязноват - дроби его более мелкие методы, выделяй какую-то логику в отдельные классы т.к. очень тяжело ревьюить - у нас метрики пока нету по величине методов, но 50 строк будто бы максимум. Нейронку если и юзаешь, то убирай её комменты хотяб

sergeJAVA commented:
Спасибо за ревью. В следующих коммитах исправлю все замечания.

gobaz095 commented on src/main/java/com/example/deal_service/controller/DealController.java:
```diff
@@ -26,8 +29,8 @@ public ResponseEntity<DealDto> getDealById(@PathVariable UUID id) {
     }
 
     @PostMapping("/save") // Для сохранения/обновления
```
Меняй комменты на документацию своих АПИ, это даже важнее чем javadoc. Популярнее всего openapi и swagger соответственно:
https://swagger.io/docs/
https://www.baeldung.com/swagger-set-example-description
https://www.youtube.com/watch?v=2o_3hjUPAfQ&pp=ygUTc3ByaW5nIGJvb3Qgc3dhZ2dlcg%3D%3D

Но помни, что javadoc != openapi. Одно для документации нашего кода, другое для документации наших api

gobaz095 commented on src/main/java/com/example/deal_service/controller/DealController.java:
```diff
@@ -45,4 +48,17 @@ public ResponseEntity<Page<DealDto>> searchDeals(
 
+    @PostMapping("/search/export")
+    public ResponseEntity<byte[]> exportDeals(@RequestBody DealSearchRequest searchRequest) {
+        byte[] excelBytes = dealService.exportDealsToExcel(searchRequest);
```
Ты говорил, что отчет пока не готов. Так как ПР отдельно идет, я не вижу реализации даже если она есть. Так что полагаю, что отчет пока не готов. Пушить еще не готовый код так себе. Ну или хотя бы используй todo

gobaz095 commented on src/main/java/com/example/deal_service/model/DealSearchRequest.java:
```diff
@@ -39,6 +39,10 @@ public class DealSearchRequest {
     private String borrowerSearch; // Для поиска по контрагентам с ролью BORROWER
     private String warrantySearch; // Для поиска по контрагентам с ролью WARRANTY
 
+    private int page;
```
Лучше сделать отдельно класс для пагинации. В прочем, в сервисе сделок у нас уже полноценный ОРМ и jpa, поэтому надо использовать их функционал.
В контрагентах без них.

gobaz095 commented on src/main/java/com/example/deal_service/model/mapper/DealMapper.java:
```diff
@@ -59,6 +60,23 @@ public static Deal toEntity(DealDto dealDto) {
+
+        Deal deal = new Deal();
+
+        deal.setDescription(request.getDescription());
```
Ты же уже знаешь вроде что такое билдер и выше использовал @Builder, почему тут не стал?

gobaz095 reviewed (commented):
Будут еще комментарии. Придется смотреть предыдущие ПРы и более подробно репозиторий. Поэтому всегда надо кидать нам на ревью ПРы, ничего не вливая самостоятельно. Целостность теряется. Я вижу в этом ПРе только обрывки приложения.

sergeJAVA committed:
refactor:

Заменены импорты с "*"; создан вспомогательный класс Pagination для пагинации, вместо полей int page, size в классе DealSearchRequest; сервис-логика в методах searchDeals и exportDealsToExcel была разбита на части (вынесено во вспомогательные классы и методы) для улучшения читабельности

sergeJAVA committed:
docs:

Добавлен javadoc; добавлена документация swagger для API; добавлен GlobalExceptionHandler для отлова кастомных исключений.

sergeJAVA committed:
docs: javadoc для ResponseException

sergeJAVA committed:
chore: добавил checkstyle

sergeJAVA committed:
test: добавил mockito тесты для API`шек.

sergeJAVA committed:
docs: добавил описание полей для Request объектов

AlKl1M commented on src/main/java/com/example/deal_service/controller/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,41 @@
+package com.example.deal_service.controller;
+
+import com.example.deal_service.exception.ContractorRoleException;
```
Все еще нейминг - почему com.example то...?

sergeJAVA replied:
Ну так сгенерировался проект на Spring Initializr, если это принципиально, то просто скажи на какой поменять

AlKl1M replied:
На Spring Initializr метадату проекта указываешь ты все таки. Ну и артифакт ты тем не менее поменял)
Не принципиально, но любой другой - хоть webbee или что-то другое. Просто example глаз режет. Ну и это не основное замечание на изменения)
![image](https://github.com/user-attachments/assets/a09ccb20-1ffa-4eb8-aa33-8635aa1f3e19)

sergeJAVA committed:
test: добавил тесты для других случаев

AlKl1M commented on src/main/java/com/example/deal_service/service/impl/ContractorRoleServiceImpl.java:
```diff
@@ -41,7 +62,7 @@ public ContractorRoleDto addRoleToContractor(ContractorRoleRequest request) {
                 return ContractorRoleMapper.toDto(link.getRole());
```
Обращаемся к полям ContractorRole, которые загружаются лениво и при массовой обработке каждый вызов getSomeField генерирует отдельный SQL-запрос. Перепроверь все подобные места еще раз - лучше избавиться от них сразу. Юзаем EntityGraph, фетчи итд. Оно тут мб не зароляет, так как я вижу тут вроде одиночная обработка, но все равно учимся юзать это.

AlKl1M replied:
п.с. я про n+1 в целом

AlKl1M commented on src/test/java/com/example/deal_service/controller/DealControllerTest.java:
```diff
@@ -0,0 +1,141 @@
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.header;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
```
В целом не конкретно к этому классу тестов, но нет возможности прицепиться к DealServiceApplicationTests, который ты уже влил в прошлом ПРе - он у тебя не работает т.к. ты зачем-то там загружаешь просто контекст. Зачем он нужен? Он падает.

Плюс мы пока не давали % на покрытие тестами, но пока не было интеграционных - резоннее тестить уровень сервиса юнит тестами. И хотя бы на 50% общего покрытия (смотреть можно через Coverage)

AlKl1M commented on src/main/java/com/example/deal_service/model/ContractorRoleRequest.java:
```diff
@@ -1,15 +1,28 @@
+/**
+ * Запрос на добавление/удаление роли у контрагента.
+ */
 @Data
```
А это не дто? Почему оно в пакете моделей?

sergeJAVA committed:
Переименовал папку example на internship; перенёс классы связанные с запросами в папку dto

sergeJAVA committed:
удалил тестовый класс, который ничего не делал, но поднимал лишний раз контекст приложения

sergeJAVA committed:
fix(DealServiceImpl): решена ошибка n+1.

Устранение N+1 в DealServiceImpl. Рефакторинг коллекций DealSum и DealContractor в Set и обновление @EntityGraph для загрузки всех связанных сущностей за один запрос.

sergeJAVA committed:
docs: updated javadoc for exportDealsToExcel method

sergeJAVA committed:
docs: moved javadoс for methods from the classes implementing an interface to their interface

gobaz095 commented:
> ![Screenshot 2025-07-01 181748](https://private-user-images.githubusercontent.com/174865702/461124740-494a45a6-9292-4885-8f24-3d06994e5fd6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE4MDI1MzgsIm5iZiI6MTc1MTgwMjIzOCwicGF0aCI6Ii8xNzQ4NjU3MDIvNDYxMTI0NzQwLTQ5NGE0NWE2LTkyOTItNDg4NS04ZjI0LTNkMDY5OTRlNWZkNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcwNlQxMTQzNThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02NWUxMjhjMTIxMDA1Y2Y5ZjIyMmIxMzA4Yjk2Y2VmYjM3YjYzM2Y0Mjc3MmUyNDg4M2M1ZWMwNThjMDg3Y2RkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.1xN8iXSSLu8zBKUlTvgQosnO_skhu7tcvRfqMxt50Sw)
> 
> получилось вот так

Покраску для строк сделай, чтобы прямо хорошо

gobaz095 reviewed (commented):
В понедельник-вторник еще буду смотреть сами репозитории целиком, так как были влиты ПРы без апрувов. Но пока критических замечаний нет.

sergeJAVA committed:
feat:

Добавил покраску для строк; добавил возможность изменять DealSum для существующих сделок (раньше к существующим просто новые добавлялись), если в запросе передаются суммы.

sergeJAVA commented:
![Screenshot 2025-07-06 184342](https://github.com/user-attachments/assets/2b8bc2b1-4c5f-42b0-8ae0-06b30b3f4600)

gobaz095 reviewed (approved):
Можно слить
